### PR TITLE
Compile Prim nodes to Hugr DFG

### PR DESCRIPTION
**Pull request by acl-cqc**
* The first couple of `examples/`files where I looked to put this seemed misnamed, so some driveby cleanups
* The test I've added is monomorphic, as anything polymorphic fails validation due to type erasure (#18)....and this seems like one where we really want validation to succeed to tell us the rest is right :)

### CODE REVIEW

File: brat/examples/vec_map.brat [hunk 1]
```diff
@@ -0,0 +1,15 @@
+ext "to_float" to_float(i :: Int) -> Float
```
Comment (acl-cqc) [inline brat/examples/vec_map.brat:1]: maybe this file would be better called "mono_vec_map.brat" (or monomorphic...)

File: brat/Brat/Compile/Hugr.hs [hunk 2]
```diff
@@ -387,8 +387,18 @@ compileWithInputs parent name = gets compiled <&> M.lookup name >>= \case
...
+    Prim (ext,op) -> do
+      let n = ext ++ ('_':op)
+      let [] = ins
+      let [(_, VFun Braty cty)] = outs
+      box_sig@(FunctionType inputTys outputTys) <- body <$> compileSig Braty cty
+      ((p, _ty), ()) <- compileConstDfg parent n box_sig $ \dfg_id -> do
+        ins <- addNodeWithInputs ("Inputs" ++ n) (OpIn (InputNode dfg_id inputTys)) [] inputTys
+        outs <- addNodeWithInputs n (OpCustom (CustomOp dfg_id ext op box_sig [])) ins outputTys
+        addNodeWithInputs ("Outputs" ++ n) (OpOut (OutputNode dfg_id outputTys)) outs [] >>= \case
+          [] -> pure ()
```
Comment (croyzor) [inline brat/Brat/Compile/Hugr.hs:399]: I think the test that there's no outputs seems superfluous (+ haskell will emit a warning about the incomplete pattern match)

Comment (croyzor) [reply brat/Brat/Compile/Hugr.hs:399]: Well, not superfluous exactly, but I don't think we need to be quite so defensive

File: brat/Brat/Compile/Hugr.hs [hunk 3]
```diff
@@ -387,8 +387,18 @@ compileWithInputs parent name = gets compiled <&> M.lookup name >>= \case
...
+      case p of
+        (Port loadConst 0) -> pure $ default_edges loadConst
```
Comment (croyzor) [inline brat/Brat/Compile/Hugr.hs:401]: Just for the sake of the GHC warning, we should add a catch all that throws an error

File: brat/test/Test/Compile/Hugr.hs [hunk 4]
```diff
@@ -50,12 +50,12 @@ nonCompilingExamples = (expectedCheckingFails ++ expectedParsingFails ++
   ,"kernel-syntax"
   ,"kinds"
   ,"let"
-  ,"list"
   ,"listpair"
   ,"one"
   ,"patterns"
   ,"qft"
   ,"test"
+  ,"type_alias"
```
Comment (croyzor) [inline brat/test/Test/Compile/Hugr.hs:58]: What changed with type alias?

File: brat/test/Test/Compile/Hugr.hs [hunk 5]
```diff
@@ -50,12 +50,12 @@ nonCompilingExamples = (expectedCheckingFails ++ expectedParsingFails ++
   ,"kernel-syntax"
   ,"kinds"
   ,"let"
-  ,"list"
```
Comment (croyzor) [inline brat/test/Test/Compile/Hugr.hs:53]: 🎉

File: brat/examples/vec_map.brat (hunk 1 above)
Comment (croyzor) [reply brat/examples/vec_map.brat:1]: Also "list" rather than "vec"

**Review by croyzor — APPROVED** [review]

File: brat/Brat/Compile/Hugr.hs (hunk 2 above)
Comment (acl-cqc) [reply brat/Brat/Compile/Hugr.hs]: Yeah, this is pretty much the same thing as your other comment - it does seem dang hard to write what are basically assertions, so instead we just drop the value and lose a (small in terms of source code) check that things are working the way we expect. And the warnings against incomplete matches (there are loads in this file!) just require us to write much longer/more verbose code to turn one kind of runtime failure into another...

What's the right thing to do here? Can we `instance MonadFail Checking` and handle all these incomplete matches in our own code (which might still just `error` but at least it'd silence the warnings)? For another PR, obviously - I've shortened/silenced these here for the time being.

**Pull request merged**